### ISSUE TITLE
Closures fundamentally broken

### ISSUE THREAD

**Issue by mathijshenquet** [jeremyletang/rgtk#28]
Closures right now are fundamentally broken:

```
fn main(){
    make_window();
    gtk::main();
}

fn make_window(){

    let mut window = gtk::Window::new(gtk::window_type::TopLevel).unwrap();
    let drawing_area = DrawingArea::new().unwrap();

    drawing_area.connect(signals::Draw::new(|cr| 
        drawing_area.get_allocated_area() //`drawing_area` is garbage
    ));

    window.add(drawing_area);
    window.show_all();
}
```

This is because rust only has stack closures for now. When `make_window` is finished and it's stack goes out of scope all referenced variables inside the callback become garbage.

I'll be able to fix this when https://github.com/rust-lang/rfcs/pull/114 gets merged.

**Comment by gkoz** [comment]
Here's my take on closures/signals: https://gist.github.com/gkoz/52ae8de5370ba99d9795
GTK wants to own the closure so we box it and give the pointer to `g_signal_connect_data` as `data`. We also specify a `destroy_data` function that will take the ownership back and run the destructors.
The weak link here is 
```rust
        unsafe {
            let closure: Box<&mut FnMut($($arg_type),*) -> $ret_type> = transmute(closure);
            (*closure)($($arg_name),*)
        }
```
where we conjure a `&mut` reference to the closure in order to invoke it. I doubt it's safe...

**Comment by gkoz** [comment]
Actually we can just be conservative and use `Fn` closures: https://gist.github.com/gkoz/52ae8de5370ba99d9795
The example uses `RefCell` anyway, so it didn't even have to change. Transmuting a pointer to an aliased reference to the `Fn` closure should be safe.

**Comment by GuillaumeGomez** [comment]
I think so too, casting should be safe. I did on other binding and it worked well. I don't see any reason why it couldn't here. It's a long time issue, it would be nice to fix it once and for all.

Do you want to do it ? I don't really have time recently so i won't do it before at least a week but maybe @jeremyletang will have it ?

But before that, we need to fix `rgtk` build ! I'll take a look this week-end.

**Comment by gkoz** [comment]
Maybe it should brew for a while more. There's another issue to tackle: the library has to translate between native GTK types and user-facing ones, possibly casting `GdkEvent` to an appropriate specific variant, converting between `Gboolean` and either `bool` or some event-specific enums (e.g. `enum Propagate { Yes, No }`).

**Comment by oakes** [comment]
Is this still a work in progress? I just ran into the problem today.

**Comment by gkoz** [comment]
My plan was to do the signals after the widgets are converted to the new model. We could probably change that. Today's `draw` signal debugging session has once again proved that a translation layer is required. Some of it is already in the tree.

**Comment by gkoz** [comment]
There're other challenges to consider. We need a way to disconnect signals or let them disconnect themselves. This probably calls for storing and managing them ourselves and not just passing pointers to GLib.

**Comment by oakes** [comment]
Is there a fundamental problem with unboxed closures that requires boxed closures to be used instead? I don't know enough about Rust internals to understand what the problem is.

**Comment by gkoz** [comment]
Don't take it as a gospel but here's my understanding:

A closures needs a place to store its environment (or at least the function pointer). An 'unboxed' closure, like any automatic variable, is stored on the stack and can't outlive the function it's defined in. Putting anything in a box means allocating on the heap and passing around a pointer that is freed when the box is destroyed. This allows a closure to live longer than the function where it was defined.

Here's some info from a more authoritative source: http://smallcultfollowing.com/babysteps/blog/2014/11/26/purging-proc/

But you don't strictly need closures to make signal handlers, you should be able to use references to any functions to avoid the lifetime problem.
```rust
fn f() { println!("Hello!"); }

fn main() {
    let fun = &f;
    fun();
}
```

Except I've just tried it and got a segfault. So need to investigate further.

**Comment by oakes** [comment]
That makes sense; I'm just surprised it ends in a segfault, but I suppose that's the nature of FFI. I don't think top-level functions would work for me, because my signal handlers rely on things from the lexical scope.

**Comment by gkoz** [comment]
Are you able to move those things into the closure i.e. `move || { do_stuff; }`? For the same reasons an unboxed closure doesn't work, any references to the automatic variables would be prohibited too.

**Comment by oakes** [comment]
I need to create several signal handlers, all of which having access to the same values, so I don't think I can use `move`.

**Comment by gkoz** [comment]
How are you going you ensure those values live long enough for the references to be valid?
The example linked above uses `Rc<RefCell<_>>`.

**Comment by oakes** [comment]
I can wrap the values I intend to use in reference-counted cells, but it doesn't solve the primary issue, which is that I have several signal handlers that need them, so I can't use `move`. If I wrap them reference-counted cells without using `move`, I still get a segfault. I admit this is not familiar territory for me.

**Comment by gkoz** [comment]
`Rc` is clonable, so you move a clone:
```rust
    let ctr: Rc<RefCell<i32>> = Rc::new(RefCell::new(0));
    let ctr1 = ctr.clone();
 
    add_key_press_handler(&window, ctr.clone());
 
    NGConnect::connect::<DeleteEvent>(&window, Box::new(move |_| {
        let mut r = ctr1.borrow_mut();
        ...
```

**Comment by oakes** [comment]
Thanks for the tip, I am currently trying it out. I made a handler creation function just like yours, and cloned the Rc objects, but on the very first `borrow_mut` call, I get a panic: `RefCell<T> already borrowed`. Have you experienced that before? Note that I am not using boxed closures -- I assume you had to fork rgtk to do that.

**Comment by gkoz** [comment]
Not sure about your problem, from the panic message it would seem that the `RefMut` from the last borrow hadn't gone out of scope before you tried to borrow again.

I didn't so much fork, just rolled my own signals in a binary (it required adding a single extern fn to the FFI though, the patch inserted at the top).
https://gist.github.com/gkoz/52ae8de5370ba99d9795

**Comment by oakes** [comment]
That's the puzzling thing -- it's the very first time `borrow_mut` is called. Perhaps it's because I'm not using boxed closures, though; I'm doing `button.connect(gtk::signals::Clicked::new(&mut move || { ... }));`. I appreciate your help.

**Comment by gkoz** [comment]
I guess I'd need to see more of the code.

**Comment by oakes** [comment]
Sure, [here's](http://pastebin.com/6N4NP9Qs) the part I'm working with right now. It's using my fork of rgtk to create a `Box` containing a `VteTerminal` with several buttons corresponding to cargo commands.

**Comment by gkoz** [comment]
Well... One possible reason could be precisely the closures brokenness: we can't be sure what's inside of the `build_term` captured variable because the closure is destroyed by the time it's called. This reason could probably be eliminated if you managed to box the closure and put it in some long-living struct, possibly a global static.
Also note that you shouldn't have to put a widget in any refcounted box (or cell) because it kinda is one already, cloning it should suffice.

**Comment by oakes** [comment]
Yeah I figured that would be a problem. I may have to revisit this later, as it is proving really difficult.
